runtime/win32.c

```diff
@@ -1342,3 +1343,88 @@ value caml_win32_get_temp_path(void)
     caml_win32_sys_error(GetLastError());
   CAMLreturn(caml_copy_string_of_utf16(buf));
 }
+
+CAMLextern char_os* caml_locate_standard_library (const wchar_t *exe_name,
+                                                  const wchar_t *stdlib_default,
+                                                  wchar_t **dirname)
+{
+  if (Is_relative_dir(stdlib_default)) {
+    LPWSTR root = NULL, basename;
+    DWORD l = MAX_PATH + 1, buf_len;
+
+    do {
+      buf_len = l;
+      caml_stat_free(root);
+      root = caml_stat_alloc(buf_len * sizeof(WCHAR));
+      l = GetFullPathName(exe_name, buf_len, root, &basename);
+    } while (l >= buf_len);
+    /* It should be an Impossible Thing for exe_name (which will have been the
+       result of GetModuleFileName) to be unparsable by GetFullPathName */
+    if (l == 0) {
+      caml_stat_free(root);
+      return caml_stat_wcsdup(stdlib_default);
+    }
+
+    CAMLassert(basename != root && Is_separator(*(basename - 1)));
```

Comment:
```suggestion
    CAMLassert(basename && basename != root && Is_separator(*(basename - 1)));
```

for safety since https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamea says:

> If lpBuffer refers to a directory and not a file, lpFilePart receives zero.

edit: Actually, what would be better is to return `stdlib_default` if `basename==0`, since `basename` is used later.

Reply:
Hang on, is it worth considering the domain of the values possible here - `exe_name` _will be_ the result of `GetModuleFileName` so it cannot possibly be a directory nor - given Windows semantics - can it have been replaced by a directory in a race (because the running program can't be deleted).

So I agree it's worth the assertion checking that `basename` isn't NULL, but surely it's not worth a code path to handle that case, because it's `basename` being NULL is going to happen with a coding error (i.e. any thing outside an assertion checking this should be an unreachable code path?)

Reply:
yes, that is true.